python/extensions/pip.bzl

```diff
@@ -14,69 +14,301 @@
+    then create two different wheel repositories that are named the following.
+
+    - @@rules_python~override~pip~pip_39_websockets
+    - @@rules_python~override~pip~pip_310_websockets
+
+    And if the wheel has any other dependies subsequest wheels are created in the same fashion.
+
+    We also create a repository for the wheel alias.  We want to just use the syntax
+    'requirement("websockets")' we need to have an alias repository that is named:
+
+    - @@rules_python~override~pip~pip_websockets
+
+    This repository contains alias statements for the different wheel components (pkg, data, etc).
+    Each of those aliases has a select that resolves to a spoke repository depending on
+    the Python version.
+
+    Also we may have more than one hub as defined in a MODULES.bazel file.  So we could have multiple
+    hubs pointing to various different pip spokes.
+
+    Some other business rules notes.  A hub can only have one spoke per Python version.  We cannot
+    have a hub named "pip" that has two spokes that use the Python 3.9 interpreter.  Second
+    we cannot have the same hub name used in submodules.  The hub name has to be globally
+    unique.
+
+    This implementation reuses elements of non-bzlmod code and also reuses the first implementation
+    of pip bzlmod, but adds the capability to have multiple pip.parse calls.
+
+    Args:
+        module_ctx: module contents
+
+    """
+
+    # Used to track all the different pip hubs and the spoke pip Python
+    # versions.
+    pip_hub_map = {}
+
+    # Keeps track of all the hub's whl repos across the different versions.
+    # dict[hub, dict[whl, dict[version, str pip]]]
+    # Where hub, whl, and pip are the repo names
+    hub_whl_map = {}
+
     for mod in module_ctx.modules:
-        for attr in mod.tags.parse:
-            requrements_lock = locked_requirements_label(module_ctx, attr)
-
-            # Parse the requirements file directly in starlark to get the information
-            # needed for the whl_libary declarations below. This is needed to contain
-            # the pip_repository logic to a single module extension.
-            requirements_lock_content = module_ctx.read(requrements_lock)
-            parse_result = parse_requirements(requirements_lock_content)
-            requirements = parse_result.requirements
-            extra_pip_args = attr.extra_pip_args + parse_result.options
-
-            # Create the repository where users load the `requirement` macro. Under bzlmod
-            # this does not create the install_deps() macro.
-            pip_repository_bzlmod(
-                name = attr.name,
-                repo_name = attr.name,
-                requirements_lock = attr.requirements_lock,
-                incompatible_generate_aliases = attr.incompatible_generate_aliases,
-            )
+        for pip_attr in mod.tags.parse:
+            # We can automatically dertermine the Python version of an
```

Comment:
I'm not against inferring the version from the name per se (we're matching our internal names, so its fine), and I'm OK with this right now.

That said, I'm leaning more and more towards the idea of using `python_version` to act as the "driver" of things.
  * (Ignas's idea) We can easily map the version to the backing repository and interpreter. Now users don't need to specify the interpreter target or use the interpreter extension.
  * With Phil's work to uncouple the pip dependency processing from the interpreter used, we would no longer need the interpreter, but would still need the python version.
  * Allowing multi-version requirements files using env markers requires us to know the intended python version; knowing the interpreter doesn't really help.

Anyways, fine for now. I filed https://github.com/bazelbuild/rules_python/issues/1268 for more thoughts here.

Reply:
> I'm not against inferring the version from the name per se (we're matching our internal names, so its fine), and I'm OK with this right now.
> 
> That said, I'm leaning more and more towards the idea of using `python_version` to act as the "driver" of things.
> 
> * (Ignas's idea) We can easily map the version to the backing repository and interpreter. Now users don't need to specify the interpreter target or use the interpreter extension.
> * With Phil's work to uncouple the pip dependency processing from the interpreter used, we would no longer need the interpreter, but would still need the python version.
> * Allowing multi-version requirements files using env markers requires us to know the intended python version; knowing the interpreter doesn't really help.
> 
> Anyways, fine for now. I filed #1268 for more thoughts here.

Not having the interpreter repo may cause some challenges with the dependencies. We would need to test it. We could also write out the sym link in the python_aliases repo. 

There are use cases where the user won’t want an interpreter, so I am wondering how we do that.

Reply:
> Not having the interpreter repo may cause some challenges with the dependencies.

How so?

> There are use cases where the user won’t want an interpreter, so I am wondering how we do that.

I'm guessing you mean a user wanting to use a local system python? The solution for that is for the user to specify the interpreter using the pip.parse `python_interpreter` arg (this accepts a path to an arbitrary interpreter).

If you mean something else, I'm not sure what you mean -- an interpreter is required in some fashion because we have to eventually run various commands using it.

Reply:
I don't follow, but lets continue discussion on #1268